ArjunaJTS/jtax/tests/classes/com/hp/mwtests/ts/jta/jts/xa/JTSTest.java

```diff
@@ -300,6 +302,39 @@ public void rollback(Xid xid) throws XAException {
         }
     }
 
+    @Test
+    public void testRollbackCalledWhenEndHasXA_RBINTEGRITY() throws Exception {
+        jakarta.transaction.TransactionManager tm = com.arjuna.ats.jta.TransactionManager
+                .transactionManager();
+
+        tm.begin();
+
+        jakarta.transaction.Transaction theTransaction = tm.getTransaction();
+
+        assertTrue(theTransaction.enlistResource(new SimpleXAResource(){}));
+
+        assertTrue(theTransaction.enlistResource(new SimpleXAResource() {
+            @Override
+            public void end(Xid xid, int flags) throws XAException {
+                assertTrue(flags == XAResource.TMSUCCESS);
+                throw new XAException(XAException.XA_RBINTEGRITY);
+            }
+
+            @Override
+            public void rollback(Xid xid) throws XAException {
+                resource1Rollback = true;
+                throw new XAException(XAException.XAER_NOTA);
+            }
+        }));
+
+        try {
+            tm.commit();
+            fail("Should not have committed");
+        } catch (RollbackException e) {
+            assertTrue(resource1Rollback);
```

Comment:
WDYT about checking the XA errorCode as well?

Reply:
Same conversation above

Reply:
This conversation will fork a bit now :)

For JTS we don't have suppressed exception support so I can't add it here